### ISSUE TITLE
Make it easier to include Community Modules

### ISSUE THREAD

**Issue by lemmy** [alygin/vscode-tlaplus#123]
Including the [TLA+ Community Modules](https://github.com/tlaplus/CommunityModules) requires one to change the global extension settings.  This is (i) cumbersome and (ii) generally not a global but a per-spec setting.

**Comment by lemmy** [comment]
Related, TLC's info statement about module overrides incorrectly shows up as an error.

![Screenshot from 2019-11-23 10-06-54](https://user-images.githubusercontent.com/88777/69483202-1df6cf80-0dd9-11ea-999c-f9ee6d1206d5.png)

**Comment by alygin** [comment]
@lemmy, you're right, it's not very flexible at the moment. There's a VSCode-way to provide folder specific configuration properties, I'll use it.

As for the wrong message level (error instead of info), I've move it to a separate issue (#124). Thanks for noticing.

**Comment by lemmy** [comment]
If CommunityModules.jar is in the current directory, the following command-line will load it and its Java module overrides automatically (on macOS & Linux):

```bash
java -cp /path/to/tla2tools.jar:CommunityModules.jar tlc2.TLC MC.tla
```

**Comment by lemmy** [comment]
One more thing to note:  None of the existing modules - e.g. the TLA+ examples - *parse* when they have a TLAPS proof and there is no way for the user to add TLAPS.tla to the library path.

**Comment by lemmy** [comment]
I feel stupid but I don't remember how I tweaked the settings to load CommunityModules.jar.  When I set ```-cp CommunityModules.jar``` as the extension's Java option, the cmd line ends up as ```java -cp CommunityModules.jar -XX:+UseParallelGC -cp /home/markus/.vscode/extensions/alygin.vscode-tlaplus-1.3.2/tools/tla2tools.jar tlc2.TLC BlockingQueueTrace.tla -tool -modelcheck -coverage 1 -config BlockingQueueTrace.cfg```

**Comment by alygin** [comment]
The only possible trick I can think of is to use Java 8 and place `CommunitiyModules.jar` in the `lib/ext` subdirectory of your Java Home. Did you do that?

**Comment by lemmy** [comment]
Would it be possible to add a new setting to add (or replace #134 ) libraries to the classpath?

**Comment by alygin** [comment]
Sure, it'll be implemented in v1.4.

**Comment by lemmy** [comment]
Does it already have a release scheduled?

**Comment by alygin** [comment]
The plan was to release it earlier in January, but unfortunately, I've been a bit overloaded at my work lately. So, it'll probably be next week. This weekend at best.

**Comment by alygin** [comment]
Available in the latest pre-release. Just provide a custom classpath in the "Java Options" setting, and the extension will merge it with the default one when running TLA commands. Providing a custom tla2tools.jar also works.

**Comment by lemmy** [comment]
How does the "merge" work?

**Comment by alygin** [comment]
I've described this option in wiki in more details: [Java Options](https://github.com/alygin/vscode-tlaplus/wiki/Java-Options)

**Comment by lemmy** [comment]
I suggest to prepend instead of append additional dependencies to tla2tools.jar in the path.  That way, a user can replace individual classes in tla2tools without replacing the complete jar. Use-case: provide Java overrides for the built-in standard modules such as Naturals, ...

**Comment by alygin** [comment]
I think adding the default path to the end is the way to go.

**Comment by alygin** [comment]
Available in the latest pre-release.

**Comment by alygin** [comment]
Released in v1.4

**Comment by ibalajiarun** [comment]
> One more thing to note: None of the existing modules - e.g. the TLA+ examples - _parse_ when they have a TLAPS proof and there is no way for the user to add TLAPS.tla to the library path. Only workaround is to copy&paste TLAPS.tla out of the TLAPS distribution into the current spec directory.

@lemmy @alygin has this particular problem been addressed? or is it expected that the user use the Options setting?
